[JS sample] Add CoreBot with Application-Insights

gasper-az opened the pull request:
## Proposed Changes
  - Add sample `21 - CoreBot with Application-Insights`, based on JS [Sample#13](https://github.com/microsoft/BotBuilder-Samples/tree/master/samples/javascript_nodejs/13.core-bot)

## More Information
This sample:
- Uses [LUIS](https://www.luis.ai) to implement core AI capabilities
- Implements a multi-turn conversation using Dialogs
- Handles user interruptions for such things as `Help` or `Cancel`
- Prompts for and validates requests for information from the user
- Uses [Application Insights](https://docs.microsoft.com/azure/azure-monitor/app/cloudservices) to monitor your bot

## Testing
To test this sample:
1. Clone this branch
1. In a terminal, navigate to `samples/javascript_nodejs/13.core-bot`
1. Install modules with `npm install`
1. Setup LUIS
1. Run the sample with `npm start`

gasper-az committed:
[AppInsights Sample] Init project [WIP]

Aliandi committed:
add ErrorViewModel class

Aliandi committed:
add cancel dialog

Aliandi committed:
add dateResolver dialog

Aliandi committed:
Change ! to cast as boolean

gasper-az committed:
[AppInsights sample] Add dialogBot

gasper-az committed:
[AppInsights sample] Export DialogBot

gasper-az committed:
[AppInsights Sample] create DialogState

gasper-az committed:
[AppInsight sample] dialogAndWelcomeBot + resource

gasper-az committed:
[AppInsight Bot] Fix styling.

Aliandi committed:
add header

gasper-az committed:
[AppInsights-Bot] Add ErrorViewModel class

gasper-az committed:
[AppInsights Bot] Add Bot folder

Add DialogBot
Add DialogAndWelcomeBot
Add Resources for DialogAndWelcomeBot

Aliandi committed:
add bookingDialog

gasper-az committed:
[AppInsights Bot] Add mainDialog [WIP]

gasper-az committed:
[AppInsights Bot] Add new Dialogs

Add BookingDialog
Add cancelAndHelpDialog
Add dateResolverDialog

gasper-az committed:
[AppInsights Bot] Add mainDialog

gasper-az committed:
[AppInsights Bot] Add MainDialog

gasper-az committed:
[AppInsights Bot] Add index.js and flightBookingRecognizer

gasper-az committed:
[AppInsight Bot] fix minor issues in dialogs

gasper-az committed:
[AppInsight Bot] Fix minor issues in dialogBot

gasper-az committed:
[AppInsights] add index.js

gasper-az committed:
[AppInsights Bot] Add cognitiveModels

gasper-az committed:
Add deployment templates

gasper-az committed:
[ApiInsights Bot] Add index.js and folders

Add Index.js
Add cognitivesModels folder
Add deploymentTemplates folder
Fix minor issues in dialog & dialogBot

gasper-az committed:
[AppInsights Bot] Add README.md file

Aliandi commented on samples/javascript_nodejs/21.corebot-app-insights/bots/dialogAndWelcomeBot.js:
```diff
@@ -0,0 +1,34 @@
+        this.onMembersAdded(async (turnContext, next) => {
+            const membersAdded = turnContext.activity.membersAdded;
+
+            for (let pos = 0; pos < membersAdded.length; pos++) {
```
Couldn't this be done with a `foreach`?

gasper-az replied:
Yes, but AFAIK, the `for` loop is more performant than the `foreach` loop ([src](https://stackoverflow.com/a/43032526))

ceciliaavila commented on samples/javascript_nodejs/21.corebot-app-insights/dialogs/mainDialog.js:
```diff
@@ -0,0 +1,173 @@
+
+    /**
+     * Shows a warning if the requested From or To cities are recognized as entities but they are not in the Airport entity list.
+     * In some cases LUIS will recognize the From and To composite entities as a valid cities but the From and To Airport values
```
Please remove the 'a' in 

> LUIS will recognize the From and To composite entities as **a** valid cities

gasper-az replied:
Typo fixed

gasper-az committed:
Sample#21 - fix typo in mainDialog.js

gasper-az committed:
Update README.md file

* Replace appsettings.json with .env file.

* Add information about appInsights instrumentation key.

gasper-az committed:
update package.json:
* Add start command
* Add watch command
* Add description

gasper-az committed:
Add source files and update package.json

*Add eslint config file

*Add gitignore file

Update package.json for adding ESLint

gasper-az committed:
Fix ESLint Errors

gasper-az committed:
bookingDialog as parameter

Add bookingDialog as a parameter of mainDialog instead of creating it.

gasper-az committed:
Apply feedback on Sample 21

- index.js: uncommented line for using restify.plugins.bodyParser.

- welcomeCard.json: set isSubtle to true.

- README.md: replace references to sample 13 with references to Sample 21.

- README.md:  add link to AppInsights documentation.

gasper-az committed:
Sample21: Use TelemetryClient through ComponentDialog

- As the mainDialog extends ComponentDialog, it s not necessary to pass the TelemetryClient as a parameter in its constructor. Instead, we can use the telemetryClient property from the ComponentDialog.

gasper-az committed:
Sample 21: Restored previous Dialog constructors.

- Restore MainDialog previous constructor.

- Restore Booking Dialog previous contructor.

- Removed unnecessary brackets.
